core/src/apps/cardano/address.py

```diff
@@ -68,22 +81,66 @@ def _has_shelley_address_header(address: bytes) -> bool:
     )
 
 
+def _validate_output_shelley_address(
+    address_str: str, address_bytes: bytes, network_id: int
+) -> None:
+    address_type = _get_address_type(address_bytes)
+    if address_type == CardanoAddressType.REWARD:
+        raise INVALID_ADDRESS
+
+    _validate_output_address_bech32_hrp(address_str, address_type, network_id)
+    _validate_address_network_id(address_bytes, network_id)
+
+
+def _validate_output_address_bech32_hrp(
+    address_str: str, address_type: int, network_id: int
+) -> None:
+    valid_hrp = _get_bech32_hrp_for_address(address_type, network_id)
+    bech32_hrp = bech32.get_hrp(address_str)
+
+    if valid_hrp != bech32_hrp:
+        raise INVALID_ADDRESS
+
+
+def _get_bech32_hrp_for_address(address_type: int, network_id: int) -> str:
```

Comment:
`_for_address` suffix seems redundant here taking into account it's just an internal function